### ISSUE TITLE
v9.6.5 EM not firing for some hooks if previous hooks resulted in an email going out

### ISSUE THREAD

**Issue by 123andy** [vanderbilt/redcap-external-modules#257]
We have a 'redcap_save_record' hook that stopped firing at 9.6.5.  We put in a bunch of debugs and got this far:

We have around 80 EMs on the server.
A particular project has 10 EMs enabled.
When we save a survey on this project, we are finding that a hook isn't working anymore.

1) When we get to the save_record hook, the External Module builds an array of all the EMs in `callHook` with `$versionsByPrefix`.  In our case we have 10 some EMs enabled.

2) It looks at each enabled EM for a save_record function and, in our 8th em finds one which has a saveData call in it.

3) When saveData is done, this triggers Alerts and Notifications going off, which triggers an outgoing email.

4) The outgoing email results in the hook 'redcap_email' being processed against ALL EMs on the server.   So, we were 8/10 through the save_record hook and now we start 1/80 on the redcap_email hook...

5) When the email hook is finished, it returns to complete the last two modules with 'save_record'.  However, a static variable inside the static ExternalModule class no longer has the right value for `self::$hookBeingExecuted` -- it is null.

So, it seems that if we keep a static class to run EMs, we need to retain the previous context before starting to process the nested email hook (or something like that..)

If we turn off alerts and notifications so that no emails are sent, then things work as expected and it processes all the hooks.  So, the injection of the redcap_email hook processing looses the information about the previous hook being run...

```
// HERE WE START PROCESSING THE REDCAP_SAVE_RECORD HOOK...
2020-02-11 14:54:59	865.9	15793	[survey respondent]	Hooks	39	call	[1/1]	STR	======= redcap_save_record =======
2020-02-11 14:54:59	866.6	15793	[survey respondent]	ExternalModules	2584	callHook	[1/1]	STR	save_record: 10
2020-02-11 14:54:59	866.7	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [1] checking stanford_webauth / v9.9.9 ...
2020-02-11 14:54:59	866.9	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [2] checking imagemap / v9.9.9 ...
2020-02-11 14:54:59	867.2	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [3] checking em_logger / v9.9.9 ...
2020-02-11 14:54:59	867.4	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [4] checking biocatalyst_link / v9.9.9 ...
2020-02-11 14:54:59	867.6	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [5] checking username_verification / v9.9.9 ...
2020-02-11 14:54:59	867.8	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [6] checking session_checker / v9.9.9 ...
2020-02-11 14:54:59	868	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [7] checking shibboleth_esignature / v9.9.9 ...
2020-02-11 14:54:59	868.1	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [8] checking proj_fac_onboard / v9.9.9 ...
2020-02-11 14:54:59	868.2	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[1/4]	STR	proj_fac_onboard
2020-02-11 14:54:59	868.2	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[2/4]	STR	v9.9.9
2020-02-11 14:54:59	868.2	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[3/4]	ARR	Array
(
    [0] => 15793
    [1] => 26
    [2] => onboarding_process_initiation_form
    [3] => 99180
    [4] => 
    [5] => F8DHTT8KNE
    [6] => 3602594
    [7] => 1
)

2020-02-11 14:54:59	868.3	15793	[survey respondent]	ExternalModules	2404	startHook	[1/1]	STR	startHook
2020-02-11 14:54:59	868.3	15793	[survey respondent]	ExternalModules	2423	startHook	[1/4]	STR	hook_save_record
2020-02-11 14:54:59	868.3	15793	[survey respondent]	ExternalModules	2423	startHook	[2/4]	STR	save_record
2020-02-11 14:54:59	868.3	15793	[survey respondent]	ExternalModules	2423	startHook	[3/4]	ARR	Array
(
    [0] => redcap_save_record
    [1] => hook_save_record
)

2020-02-11 14:54:59	868.3	15793	[survey respondent]	ExternalModules	2423	startHook	[4/4]	STR	26
2020-02-11 14:54:59	868.9	15793	[survey respondent]	ExternalModules	2461	startHook	[1/1]	STR	in StartHook activeModulePrefix : proj_fac_onboard / recID: 26 / result :  / hookName : save_record


// HERE MID WAY THROUGH SAVE_RECORD FOR EM8 WE PROCESS ALL EMS FOR REDCAP_EMAIL
2020-02-11 14:54:59	953.7	15793	[survey respondent]	Hooks	39	call	[1/1]	STR	======= redcap_email =======
2020-02-11 14:54:59	953.8	15793	[survey respondent]	ExternalModules	2584	callHook	[1/1]	STR	email: 83
2020-02-11 14:54:59	953.9	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [1] checking mass_delete / v9.9.9 ...
2020-02-11 14:54:59	954.1	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [2] checking mirror_master_data / v9.9.9 ...
2020-02-11 14:54:59	954.3	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [3] checking shazam / v9.9.9 ...
2020-02-11 14:54:59	954.6	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [4] checking autocontinue_logic / v9.9.9 ...
2020-02-11 14:54:59	954.8	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [5] checking survey_ui_tweaks / v9.9.9 ...
																										 ...
2020-02-11 14:54:59	971.1	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [55] checking session_checker / v9.9.9 ...
2020-02-11 14:54:59	971.3	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [56] checking survey_portal / v9.9.9 ...
2020-02-11 14:54:59	971.6	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [57] checking proj_fac_onboard / v9.9.9 ...
2020-02-11 14:54:59	971.7	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[1/4]	STR	proj_fac_onboard
2020-02-11 14:54:59	971.7	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[2/4]	STR	v9.9.9
2020-02-11 14:54:59	971.7	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[3/4]	ARR	Array
(
    [0] => jael@stanford.edu
    [1] => faculty_onboarding_system@lists.stanford.edu
    [2] => ACTION REQUIRED : Paper Rock Faculty Onboarding Initiation Form Review
    [3] => <p><img src="https://redcap.stanford.edu/plugins/open/resources/stanford-medicine.png" alt="Stanford Medicine" style="width:350px;height:50px;"></p>
<p> </p>
<p>Dear Department of Anesthesia,</p>
<p>Andrew B Martin has requested approval to begin the faculty onboarding process for Paper Rock. </p>
<p><a href=https://redcap-dev.stanford.edu/surveys/?s=mxDcaqM8uI>Click here to review and approve this request</a></p>
<p><i>Note: the onboarding process will not move forward until the approval has been granted.</i></p>
    [4] => 
    [5] => 
    [6] => 
    [7] => Array
        (
        )

)

2020-02-11 14:54:59	971.7	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[4/4]	ARR	()
																										 ...
2020-02-11 14:54:59	977.3	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [80] checking parent_child / v9.9.9 ...
2020-02-11 14:54:59	977.5	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [81] checking biocatalyst_link / v9.9.9 ...
2020-02-11 14:54:59	977.7	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [82] checking giftcard_reward / v9.9.9 ...
2020-02-11 14:54:59	977.9	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [83] checking record_naming / v9.9.9 ...

// HERE WE RESUME PROCESSING THE LAST TWO SAVE_RECORD EMs
2020-02-11 14:55:01	2198.6	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [9] checking ssn_masker / v9.9.9 ...
2020-02-11 14:55:01	2198.8	15793	[survey respondent]	ExternalModules	2590	callHook	[1/1]	STR	 [10] checking summarize / v9.9.9 ...
2020-02-11 14:55:01	2198.9	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[1/4]	STR	summarize
2020-02-11 14:55:01	2198.9	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[2/4]	STR	v9.9.9
2020-02-11 14:55:01	2198.9	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[3/4]	ARR	Array
(
    [0] => 15793
    [1] => 26
    [2] => onboarding_process_initiation_form
    [3] => 99180
    [4] => 
    [5] => F8DHTT8KNE
    [6] => 3602594
    [7] => 1
)

2020-02-11 14:55:01	2198.9	15793	[survey respondent]	ExternalModules	2568	ExternalModules\{closure}	[4/4]	ARR	()
2020-02-11 14:55:01	2199	15793	[survey respondent]	ExternalModules	2404	startHook	[1/1]	STR	startHook
2020-02-11 14:55:01	2199	15793	[survey respondent]	ExternalModules	2423	startHook	[1/4]	STR							// HOWEVER, HERE OUR self::$hookBeingExecuted IS NULL...
2020-02-11 14:55:01	2199	15793	[survey respondent]	ExternalModules	2423	startHook	[2/4]	BOOL	false
2020-02-11 14:55:01	2199	15793	[survey respondent]	ExternalModules	2423	startHook	[3/4]	ARR	Array
(
    [0] => redcap_
    [1] => hook_
)

2020-02-11 14:55:01	2199	15793	[survey respondent]	ExternalModules	2423	startHook	[4/4]	NULL	
2020-02-11 14:55:01	2199.3	15793	[survey respondent]	ExternalModules	2431	startHook	[1/1]	STR	noPermission
2020-02-11 14:55:01	2230.9	15793	[survey respondent]	Hooks	39	call	[1/1]	STR	======= redcap_every_page_top =======

```

**Comment by 123andy** [comment]
Found a hack to make it work again, but I think the use of static variables in the EM class when it can be called 'recursively' is problematic...

```
$startHook = function($prefix, $version) use ($arguments, &$resultsByPrefix){
    $before = self::$hookBeingExecuted;
 
    $result = self::startHook($prefix, $version, $arguments);

    if ($before !== self::$hookBeingExecuted) self::$hookBeingExecuted = $before;
```

**Comment by mmcev106** [comment]
@123andy, I wanted to let you know this is definitely on my list to look into.  I'm not sure when I'll be able to prioritize it just yet.

**Comment by 123andy** [comment]
Thanks for letting me know you've seen the issue.  It is a pretty big bug that many will not discover as it just doesn't execute some of the hooks in certain conditions... somewhat hard to replicate.  The main problem is storing a changing 'state' in the static EM class...   I would consider it a high-priority.

**Comment by grezniczek** [comment]
Maybe there are some synergies with addressing #261, as some (new) state management will be needed for a proper solution there as well.

**Comment by 123andy** [comment]
Bump - I'd love to get this resolved and see your input...  I think the issue is that any context, such as the current EM - probably shouldn't be stored in the actual external module object - or, if so, it needs to be carefully managed.  I'm guessing we are not the only users that will be seeing bugs due to this and not realizing it.

**Comment by grezniczek** [comment]
Would it make sense to implement hook processing, or rather the triggers that initiate it, through a queue? Queued triggers will be processed one by one, as they were added to the queue. This would prevent starting new processing chains in the middle of previous ones.
Now I haven't really thought this through in very careful detail yet, but are there any downsides to this approach? 

I'd be willing to (help) work on an implementation, but we'd have to agree on a (somewhat detailed) strategy first.

**Comment by 123andy** [comment]
Are there any other cases where, inside of a hook method, you call another hook in the same process?  I can't think of any...  

So, what you're suggesting is that email hooks be handled differently and queued up until the end? 
 I'm not sure that I follow how that would work... 

In this case, when processing the save_record hook, an alert and notification is triggered.  This then results in an email going out which results in another hook being evaluated...  How could that process be queued?

It seems to be that when we 'process a hook' we may want to instantiate an object based on the ExternalModule class to manage that particular hook method.  This instance would contain the name of the method being run and track its progress, rather than piggy-back on the static ExternalModule class...    In this case, the 'save_record' hook method would instantiate an \ExternalModuleRunner or something that then processes the save_record hook.  In the middle, when an Alert is triggered, another instance of the Runner would be created to parse the redcap_email hook...

**Comment by grezniczek** [comment]
The goal would be to "serialize" hook processing within a request. Instead of "interrupts" when processing a hook triggers another, I'd propose to queue it and only start processing after all active modules had the first hook processed. 
So let's say, during 'save_record' processing, module 5 of 7 sends an email, thus triggering the email hook, the 'save_record' gets finished first (for modules 6 and 7), and only then, the 'email' processing starts (for all modules that implement the email hook).

Now, I haven't really thought about a concrete implementation, but I am sure such a queue within a single request is a problem that has been solved. Is concurrency even a problem? Are static members of classes isolated between different requests? I think so, as each request is an idependent bootstrap of PHP, isn't it?

How could the process be queued? Assuming there are no concurrency issues, this could be a simple array. New hook triggers are added to the end. A processor works its way from start to finish.

Or am I thinking in a totally wrong (=can't ever work) direction?

**Comment by mmcev106** [comment]
[This recent PR](https://github.com/vanderbilt/redcap-external-modules/pull/283) to prevent an infinite loop with email hooks helped me get into the game on this one.  Apart from emails, I don't know of any other expected cases for nested hooks.  I added an error log in that PR that will notify us of other cases so we can consider them.

I have two ideas for a solution to this issue that I could implement now:

1. Change `$hookBeingExecuted` to an array, which would be used to track the stack of nested hook calls.
2. Pass the hook name to the `startHook()` function, instead of using `$hookBeingExecuted` to determine which hook should be called  (`$hookBeingExecuted` would still be needed for error handling).

I might wait to decide on which option makes more sense until I get into the details (especially around error handling).

@123andy & @grezniczek, do you have any concerns or alternate recommendations before I start work on one or both of those options?  I'll plan to wait to get your thumbs ups before I start.

**Comment by 123andy** [comment]
Thanks for looking into this, @mmcev106 .     Regarding recursive hooks - could it ever be the case that we would allow the programmatic 'triggering' of a hook as part of some other process?  For example, imagine someone wrote a data import utility and after importing each record, they wanted to have an option to call the 'redcap_save_data' hook on the imported data to add some calcs or do something else?  Then, could you imagine that the importer tool could be packaged as an EM and triggered via a cron.

In such a case you would have a cron process that runs an EM1.  EM1 then imports a record into a project and triggers a redcap_save_data which might trigger a second EM, EM2.  Then, inside of this EM2 redcap_save_data hook, an email is triggered due to an alert and notification, and you have an email handler hook, so you get a third level nested hook, EM3.

It would be good if we could make a solution that could handle this in the future...  (and probably detects if it is ever caught in an infinite loop... -  or maximum depth...)

I like your idea - basically the `$hook_being_executed` would be a register where you add and remove to the stack and also pass the current context into the start_hook?    

Perhaps related - I recall trying to debug some issues in the past that I think were related to this.  My emLogger EM is used as a 'utility' class of sorts, where, inside of many other EMs, I grab an instance of the emLogger and ask it to do its thing.  There was a bug that I couldn't quite figure out that had to do with the ExternalModule class having some static variables that were getting confused and ending up null so it didn't know what the current module was....  or, something like that...

**Comment by grezniczek** [comment]
@123andy @mmcev106 
Wouldn't the "queue approach" solve this in a more robust and future-proof way? It would cover the scenario of programmatically triggering specific hooks easily. Also, it would ensure some level of reproducibility, in that hook implementations in EMs are always executed in a predictable order. Certainly a benefit. With the current "interrupt" style, this can get very unpredicatable.

Infinite loops still are an issue and need to be broken if they happen. There are probably a few ways to identify them, maybe by keeping track of the sequence of executed hooks and ringing the alarm bells as soon as a sequence at the end gets repeated. 
With the "queue approach", this could only ever happen after all EMs that are not part of the infinite loop have had their go, again adding to predictability, reliability, and reproducibility.

This would also provide a nice environment for special handling of (multiple active) EMs implementing the email hook (see discussion in #261), and adding mechanism that support some EMs to be "trackers" only while some are "modifiers". If modules would (have to) announce their intentions regarding this in config.json, this would allow for some powerful stuff by combining multiple EMs that implement the email hook and still have predicable (and  potentially admin-controllable) outcomes.

**Comment by 123andy** [comment]
@grezniczek - I may not be understanding your queue approach.  

Say you have 10 projects with save_record implemented.  EM 5 sends an email which triggers the email hook.  Rather than evaluate redcap_email now, you want queue the running of the email hook (and save context parameters).

EMs 6-10 finish calling save_record and THEN, the next hook on the queue is email, so it starts a run of email with the parameters saved from EM5 before...

I can see both as being interesting...  Would there be any issues with global variables? I don't think so...

But there would be an issue with return variables, right? Can you actually finish the execution of your save_record hook without having the outcome of a subsequent email hook?  Not sure...

Would the queue look like this?
```
[
 [0] => {
      "hook": "redcap_save_record",
      "args": [ "project_id": 123, "record": "22", ... ]
      },
  [1] => {
     "hook": "email",
     "args":  [to: foo, from: bar, subject: asdf]
     },
  [2] => {
     "hook": "email",
     "args":  [to: ann, from: mary, subject: qwerty]
     }
]
```
I dunno...  I'm easily confused.

I was thinking something more like this:

When a hook function is called we end up scanning EMs for this function and evaluating those functions.  This process is currently 'run' by the ExternalModule as a static class and the hook being executed is stored in a static var.

Would it not make more sense to 'instantiate' an object to run all of the EMs with that hook defined?  then, if during execution we spawn the evaluation of a hook, we span a new object to manage its execution.  This would be a recursive method.  I think what @mmcev106 is talking about is just having a register that keeps track of the current hook method which I think would also work.  

I'm good with whatever works so long as AEM's doing have issues because of state confusion being stored in the ExternalModule class.

**Comment by mmcev106** [comment]
@grezniczek, the "queue approach" is appealing, but I'm having trouble envisioning an implementation that would provide the same functionality/flexibility that we have today. 
The [discussions that resulted in the delayModuleExecution() method](https://github.com/vanderbilt/redcap-external-modules/issues/257#issuecomment-604676563) may be relevant here.

**Comment by grezniczek** [comment]
I just went through the current implementation in more detail. Please correct me if I am wrong. The main (only) entry point is `callHook` which acts as the showrunner, calling `startHook` for every module enabled on the system. The only outside influences are `delayModuleExecution` and `exitAfterHook`. And, well, the chance that a hook execution triggers a new call to `callHook`, thereby interrupting the processing (as the new call "gets priority"). 

With the _serial queue_, a call to `callHook` would only result in the call to be added to a queue (an array or some suitable structure), similar to what @123andy sketched above, but probably with some additional state information. The essentials of the remaining code in `callHook` would largely stay as is, but be moved to a new `hookRunner` function with some added logic for processing hooks as described above.  Some stuff done in startHook would probably move into `callHook`, so that the specific modules that are actually active in the given context and have the hook will be part of the queue in some manner. `delayModuleExecution` and `exitAfterHook` would be unchanged from the perspective of EMs. Their _AbstractExternalModule_ implementations, however, would probably need to provide _prefix_ and _version_ as parameters to support (or at least simplify) state management in the runner.

The only somewhat serious issue with this is getting the return values back to the callers. This would have to be implemented as a callback instead of simply returning a value. However, as far as I can see, _REDCap/Classes/Hooks.php_ is the only place where this is relevant, and this could easily be changed to a callback implementation.
I did this quick and dirty on my dev box with a brand new 9.8.0 REDCap and everythings appears to work just fine.
(Well, except that -independent of my callback hack- the extensive under-the-hood overhauls in 9.8.0 appear to have broken some functionalities of one of my bigger modules. Ouch.  It'll be a bit before I can investigate, so 9.8.0 won't go live anytime soon :( Shazam seems to be affected, too, it seems)

Here's that modified code: 
```php
// Call a REDCap Hook function
public static function call($function_name, $params=array())
{
	// If not a real hook method in this class, then return
	if (!method_exists(__CLASS__, $function_name)) return false;

	// Explicitly set the default result value if it does not get set below
	$result = false;

	// Call External Modules method before calling the native hook in the hook functions file
	if (defined("APP_PATH_EXTMOD")) {
		$GLOBALS['__currently_inside_hook'] = true; // Set flag to prevent initiating the building of the record list cache
		$result = \ExternalModules\ExternalModules::callHook($function_name, $params, null, [__CLASS__, "callCallback"]);
		unset($GLOBALS['__currently_inside_hook']); // Unset flag now that the hook has been called
	}
}
static function callCallback($function_name, $result) 
{
	// Evaluate hook from hook_functions.php file if present
	if (function_exists($function_name)) {
		// Call the hook function
		$func_result = call_user_func_array($function_name, $params);
		if ($func_result !== null) {
			// We have to decide how to handle collisions where both the EM AND the hook function return a value
			// For now, I'll let the local function override the EM as any local hook function has to be
			// explicitly set up by the owner of the redcap instance
			$result = $func_result;
		}
	}
	// Call the appropriate method to process the return values, then return anything returned by the custom function
	return call_user_func_array(__CLASS__ . '::' . $function_name, array($result));
}
```

Also, I guess the `__currently_inside_hook` global would not be needed any more, but I haven't checked. This appears to be only occuring once in _REDCap/Classes/Records.php_, and was probably needed because of the interrupty way hooks are processed now? 

In `callHook`, I put this at the end (in an if statement to facilitate simply switching back and forth between the different `Hook::call` implementations):
```php
	$rv = self::filterHookResults($resultsByPrefix, $name);
	if ($callback !== null) {
		$callback($callback_name, $rv);
	}
	else {
		return $rv;
	}
}
```
_$callback_name_ captures the hook name as it comes into `callHook`.

So it would seem a hook-processing following an linear, ordered queue (doing away with "interrupts") might be feasible without any changes to EMs.

This all might then be extended with new features, such as maybe way to specify in config.json whether a module's hook should run after a certain other EMs (by giving their prefixes), and maybe other useful options.

**Comment by mmcev106** [comment]
I think this would work.  It does add complexity when our support time is already maxed out and we have a backlog of issues though.  What would be the advantage(s) of this over the simpler solutions?  We may want to wait to discuss or implement something more complex until we have more concrete examples.

**Comment by grezniczek** [comment]
I guess the big advantage would be to get rid of the "interrupty" nature of the current implementation that seems to be at the root of many of the problems that had to be solved in the past and others that may still be unadressed or have not surfaced yet.

A huge framework (and module) developer benefit would be that step-through-debugging will become much more linear. At least in my dev env, when I step into this part, it's like a (bad) roller coaster ride. Don't you see that in your tooling?

**Comment by mmcev106** [comment]
I may need to specifics to fully understand your perspective.  A method call stack is "interrupty" by nature, but it might be easier to troubleshoot on average than a web of event based interactions.  At the moment I feel like a queued hook implementation might make it harder to trace a series of events.

**Comment by grezniczek** [comment]
As I would implement it, the series of events would be defined by the order in the queue, and this order can be maintained in a "done" array, available for inspection any time during the life cycle of the request.

After my inspection of the callHook/startHook yesterday in the attempt to understand what's going on, I feel that implementation of the queue, limited to replicate what's going on now (i.e. not adding any new features), might be easier than anticipated (I hope this holds true ;). I may be able to give it a go during the weekend. I'll keep you posted.

**Comment by grezniczek** [comment]
What do you think, @123andy ?

**Comment by mmcev106** [comment]
Two more thoughts I had before you go to too much trouble:
- If we want to support nested save hooks, a queue may not be preferred since the outer hook may want to wait for the inner too hook to complete before continuing.
- I may not be able to justify changing hook behavior without identifying more specific examples of use cases that warrant the change.

**Comment by 123andy** [comment]
Hi @grezniczek  - I think it is an interesting idea, but I'm not sure I see an upside for making a big change.  Then again, whoever is willing to spend the time to make it work probably gets some latitude...  Most events are chained so I think the current, 'recursive', flow is what I am accustomed to so switching to a function queue would take some reorientation...

**Comment by grezniczek** [comment]
I have been thinking about this some more, and came to the conclusion that @123andy 's idea of instantiated hook runners (that keep track of all necessary state) appears the best way to go.

The most compelling argument was that about save actions relying on "nested" ones getting done first. This would be very complicated if not impossible to implement with a queue approach as I had described. I think I was too much preoccupied with thinking about how the rather special email hook could/should be handled, and where a "queue approach" might be useful.

**Comment by mmcev106** [comment]
I am able to justify working on the two simpler solutions I previously posted.  I'm not sure I understand the advantage of a hook runner class at the moment (over an array of hook names), but maybe that would become apparent after getting into it.  Are either of you interested/able to work on this?  If not, do you have additional input before I start on a simpler solution?  I'll wait at least a few days to hear back before starting on anything.

**Comment by 123andy** [comment]
I just upgraded to 9.8.4 and re-applied my simple patch to remember the previous hook as there wasn't an official fix in place.  I do think this is a pretty important one to resolve.  I think @mark's approach of removing the state from the static and putting it into a queue so it can run recursively works for me...

**Comment by mmcev106** [comment]
I'm heads down on a high priority COVID project currently, and am struggling to just stay caught up with emails & PRs related to the framework.  It may be weeks before I'll realistically be able to think about this.  We have not encountered this issue at Vanderbilt yet (that we know of).  If either of you want to write a PR for this before I get to it, I'll push it through as fast as I can.

To clarify, my idea was still to use a static, but just make it an array of strings (instead of a single string), that we could effectively use as a stack.  That would be a simple change, but warrants serious testing.  I'm up for discussing more complex changes, but I think we want to simplest change that accomplishes the goal to minimize technical debt and long term support consequences.

**Comment by mmcev106** [comment]
I've apparently been mulling this over subconsciously because a simple way to do it just came to me while looking at something unrelated.  I believe PR #315 might fully address this.  I'm curious to hear your thoughts.

**Comment by 123andy** [comment]
Looks pretty similar to my post at the top of this thread :-)
https://github.com/vanderbilt/redcap-external-modules/issues/257#issuecomment-584917587

**Comment by mmcev106** [comment]
@123andy, I apologize, I thought there was reason you hadn't made a PR for that.  I was thinking it didn't cover all scenarios or something.  It looks l just fine now that I have my head wrapped around the problem.  I guess having the stack present might be useful for debugging in the future, but part of me favors the simpler solution...

**Comment by 123andy** [comment]
You don't have to apologize - thanks for all you do.  

It just reminded me of my PI in grad school..  The first time anyone would bring up an idea, he would say, "Ahhh.. yea... I don't think that's such a good idea...".  Then, after about 15 minutes of discussion, he would say, "Holy Sh*t.. Okay.. you really want to know how you should do it?  This is what you should do.." and proceed to restate what we had posited at the beginning of the conversation.  It happened so often we all came to expect it :-)

Ahh.. I needed a good laugh remembering those days...

**Comment by mmcev106** [comment]
It feels good to take credit for an idea.....regardless of who came up with it.

**Comment by mmcev106** [comment]
Upon further review, I think all previous discussion has only considered partial solutions to nested hook support.  A full solution would need to encapsulate all of the following variables into a single static `$hookState` variable/object:
- $delayed
- $delayedLastRun
- $exitAfterHook
- $hookStartTime
- $hookBeingExecuted
- $versionBeingExecuted
- $activeModulePrefix

That `$hookState` variable would then get initialized and restored to it's previous state at the beginning and end of `callHook()`, as opposed to `startHook()`.

@123andy and @grezniczek, I would very much like your input on this approach if/when you are able to consider it.

**Comment by 123andy** [comment]
I know I've seen some error messages from the EM's that indicate a schizophrenic class where half the variables are in one land and another 1/2 in the other...  So, I think removing all STATE from the static class would be great - hence the best solution is to instantiate a 'runner' if we can get there...  I'm saying all of this after a beer or two with no intention of doing it myself, mind you :-)

Have a great weekend!

**Comment by mmcev106** [comment]
If I'm interpreting that correctly, I think the 'runner' and `$hookState` are the same concept.  I was thinking it might be easiest if the reference to the current state object is still stored statically.  It might be possible to pass the hook state to each module instance instead (for use by `delayModuleExecution()`), but I'm having trouble thinking of a way to implement that without adding more complication than it's worth.

**Comment by mmcev106** [comment]
I think 87855d2 might be a good start toward a solution that works in all cases.  Every time I have an hour to spare I plan to move another one of the [static state vars](https://github.com/vanderbilt/redcap-external-modules/issues/257#issuecomment-622562848) to the `HookRunner` class.  Do either of you have any concerns about this approach before I spend too much time on it?  Thanks very much.

**Comment by 123andy** [comment]
Thank you, @mmcev106 - I think you're solution looks to be simple and effective.  I appreciate all your work on this!

**Comment by grezniczek** [comment]
I have not had a chance to look at it yet. Maybe I get to it later today, but probably only on Thursday.
@mmcev106 - regarding your comment on the [community site](https://community.projectredcap.org/comments/83997/view.html), maybe this is something that could be integrated into the `HookRunner` (note, I am saying this without having had a look at https://github.com/vanderbilt/redcap-external-modules/commit/87855d258219877fcf37995d6adb8cfed0573e00 yet). Same for the email hook, eventually, but that's gonna be a little more involved I'm afraid.

**Comment by grezniczek** [comment]
@mmcev106, is your plan to only move state to the `HookRunner` class? Or to also move the code that does the work (so basically all of startHook and any new handling of return values etc) to this class? To move all hook execution logic out of ExternalModules to a dedicated class is quite appealing. All that would remain is `callHook` that would instantiate a new `HookRunner` with all applicable state and then just do something like `$result = $hookRunnerInstance->execute();`, followed by `exit();` or `return $result;`. 

There is probably no need to even keep track of a "currentHookRunner", but to add them to a list that's maintained in static ExternalModules would be nice, as this would allow keeping track of any hooks having been run, in order, during a requests lifetime when debugging. These instances could/should maybe even keep a kind of log of what they did, and specifically the results returned by the modules' hook implementations they called.

Hope this makes sense.

**Comment by mmcev106** [comment]
That all makes perfect sense.  My current goal is to change the smallest amount of code possible to fully support nested hooks.  After that I should likely prioritize other goals that currently affect users 
 or security/stability before any optimizations.

**Comment by mmcev106** [comment]
PR #345 has been merged, so this should be fixed in an upcoming REDCap release in the next few weeks.
